### ISSUE TITLE
[LoRA Phase 10.4] Implement Mixed Precision Training Support

### ISSUE THREAD

**Issue by makr-code** [makr-code/ThemisDB#564]
## 📋 Description

**Priority**: Phase 10.4 - Priority #4 (Performance + Memory Optimization)  
**Prerequisites**: Phases 1-9 complete, kernel fusion (Phase 10.3) recommended  
**Estimated Effort**: 3 weeks  
**Status Document**: `LORA_GPU_PHASE10_PLAN.md`

Implement mixed precision training (FP16/BF16) for GPU-accelerated LoRA training. Mixed precision provides 2x speedup and 2x memory reduction while maintaining training stability through FP32 master weights.

## 🎯 Goals

- [ ] Implement FP16 (half precision) tensor operations
- [ ] Implement BF16 (bfloat16) tensor operations (CUDA/HIP)
- [ ] 2x faster training (FP16/BF16 compute)
- [ ] 2x less VRAM usage (FP16/BF16 storage)
- [ ] Maintain numerical stability with FP32 master weights
- [ ] Support automatic mixed precision (AMP) training

### 1. Mixed Precision Tensor Types
- [ ] Add FP16 data type support to GPUTensor
- [ ] Add BF16 data type support (CUDA/HIP only)
- [ ] Implement data type conversion kernels (FP32↔FP16↔BF16)
- [ ] Update tensor creation utilities for mixed precision
- [ ] Add dtype parameter to tensor constructors

**Files**:
- `include/llm/lora_framework/gpu_tensor.h` (update)
- `src/llm/lora_framework/gpu_tensor.cpp` (update)
- `include/llm/lora_framework/tensor_dtype.h` (new)

**Data Types**:
```cpp
enum class DType {
    FLOAT32,  // Full precision
    FLOAT16,  // Half precision (FP16)
    BFLOAT16  // Brain float16 (BF16)
};
```

### 2. FP16 Compute Kernels (CUDA)
- [ ] Implement FP16 matrix multiplication (cuBLAS + Tensor Cores)
- [ ] Implement FP16 element-wise operations
- [ ] Implement FP16 transpose and gradient kernels
- [ ] Use `__half` and `__half2` types for vectorization
- [ ] Enable Tensor Core acceleration (Ampere+)

**Files**:
- Update `src/llm/lora_framework/kernels/cuda_kernels.cu`
- `include/llm/lora_framework/cuda_fp16_kernels.h` (new)

**Performance**: 
- Tensor Cores: Up to 2-4x faster than FP32
- Memory bandwidth: 2x less data movement

### 3. BF16 Compute Kernels (CUDA/HIP)
- [ ] Implement BF16 matrix multiplication
- [ ] Implement BF16 element-wise operations
- [ ] Use `__nv_bfloat16` (CUDA) and equivalent (HIP)
- [ ] Enable BF16 Tensor Core support (Ampere+)

**Files**:
- Update CUDA/HIP kernel files
- `include/llm/lora_framework/cuda_bf16_kernels.h` (new)

**Benefits of BF16**:
- Same dynamic range as FP32 (8-bit exponent)
- Better numerical stability than FP16
- Native support on newer GPUs (Ampere+, MI200+)

### 4. Mixed Precision LoRA Layer
- [ ] Create `MixedPrecisionGPULoRALayer` class
- [ ] Forward pass in FP16/BF16 (compute dtype)
- [ ] Backward pass in FP16/BF16
- [ ] Maintain FP32 master weights
- [ ] Automatic loss scaling for FP16 stability
- [ ] Gradient accumulation in FP32

**Files**:
- `include/llm/lora_framework/mixed_precision_lora.h` (new)
- `src/llm/lora_framework/mixed_precision_lora.cpp` (new)

**Architecture**:
```
Input (FP32) → Cast to FP16
    ↓
Forward (FP16 compute, Tensor Cores)
    ↓
Output (FP16) → Loss (FP32 for stability)
    ↓
Backward (FP16 compute)
    ↓
Gradients (FP16) → Accumulate in FP32
    ↓
Optimizer (FP32 master weights)
```

### 5. Automatic Mixed Precision (AMP)
- [ ] Implement AMP context manager
- [ ] Dynamic loss scaling for FP16 training
- [ ] Overflow/underflow detection
- [ ] Automatic scale adjustment
- [ ] Gradient scaling and unscaling

**Files**:
- `include/llm/lora_framework/amp.h` (new)
- `src/llm/lora_framework/amp.cpp` (new)

**Loss Scaling**:
- Start with scale = 65536 (2^16)
- Increase scale by 2x every N steps (no overflow)
- Decrease scale by 2x on overflow
- Skip optimizer step on overflow

### 6. Mixed Precision Optimizer
- [ ] Update `GPUSGDOptimizer` for mixed precision
- [ ] Store master weights in FP32
- [ ] Compute updates in FP32
- [ ] Cast updated weights to FP16/BF16 for training
- [ ] Support gradient clipping in FP32

**Files**:
- Update `src/llm/lora_framework/gpu_lora_layers.cpp`

### 7. FP16/BF16 Support in HIP
- [ ] Implement FP16 kernels for AMD GPUs
- [ ] Implement BF16 kernels (MI200 series)
- [ ] Use rocBLAS mixed precision GEMM
- [ ] Optimize for RDNA2/RDNA3/CDNA architectures

**Files**:
- Update `src/llm/lora_framework/kernels/hip_kernels.cpp`

### 8. Vulkan/DirectX FP16 Support (Optional)
- [ ] Add FP16 support to Vulkan shaders
- [ ] Add FP16 support to DirectX shaders
- [ ] Use 16-bit storage buffer extension (Vulkan)
- [ ] Use min precision hints (DirectX)

**Files**:
- Update Vulkan/DirectX shader files

### 9. Testing and Validation
- [ ] Unit tests for FP16/BF16 tensor operations
- [ ] Numerical accuracy tests (mixed precision vs FP32)
- [ ] Training convergence tests (verify no degradation)
- [ ] Performance benchmarks (FP16 vs FP32)
- [ ] Memory usage validation (2x reduction)
- [ ] Overflow/underflow detection tests

**Files**:
- `tests/test_mixed_precision.cpp` (new)
- Update `tests/test_gpu_lora_layers.cpp`

### 10. Documentation
- [ ] Mixed precision training guide
- [ ] When to use FP16 vs BF16
- [ ] Troubleshooting numerical stability issues
- [ ] Performance tuning for Tensor Cores
- [ ] Update Phase 10 plan with results

## ✅ Acceptance Criteria

- [ ] FP16 and BF16 tensor operations functional
- [ ] Mixed precision training maintains accuracy (< 1% loss degradation)
- [ ] **Performance**: 2x faster training (3.2ms → 1.6ms per step)
- [ ] **Memory**: 2x VRAM reduction (10GB → 5GB for Llama-7B)
- [ ] Automatic mixed precision with loss scaling works
- [ ] FP32 master weights maintained for stability
- [ ] Tensor Core acceleration verified on supported GPUs
- [ ] All tests pass with mixed precision
- [ ] Training convergence matches FP32 baseline

## 🔗 Dependencies

- CUDA 11.8+ with cuBLAS (FP16/BF16 support)
- ROCm 5.0+ with rocBLAS (FP16/BF16 support)
- Ampere GPUs (RTX 30/40 series) for Tensor Cores
- MI200 series for AMD BF16 support
- Existing GPU backends and kernels

### Speedup (with Tensor Cores)
| Operation | FP32 Time | FP16 Time | Speedup |
|-----------|-----------|-----------|---------|
| MatMul (768×768) | 0.1ms | 0.05ms | 2x |
| Forward pass | 1.0ms | 0.5ms | 2x |
| Backward pass | 2.0ms | 1.0ms | 2x |
| **Training step** | **3.2ms** | **1.6ms** | **2x** |

### Combined with Kernel Fusion
| Optimization | Training Step | Cumulative Speedup |
|--------------|---------------|-------------------|
| Baseline (CPU) | 160ms | 1x |
| GPU (FP32) | 3.2ms | 50x |
| + Kernel Fusion | 1.75ms | 91x |
| + Mixed Precision | **0.9ms** | **178x** |

### Memory Reduction
| Model | FP32 VRAM | FP16 VRAM | Reduction |
|-------|-----------|-----------|-----------|
| Llama-7B (r=8) | 10 GB | 5 GB | 2x |
| Llama-13B (r=8) | 16 GB | 8 GB | 2x |
| Llama-30B (r=8) | 32 GB | 16 GB | 2x |

## 📚 References

- NVIDIA Mixed Precision Training: https://docs.nvidia.com/deeplearning/performance/mixed-precision-training/
- FP16 vs BF16: https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus
- CUDA FP16 Programming: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#half-precision
- Phase 10 Plan: `LORA_GPU_PHASE10_PLAN.md`

### Numerical Stability Considerations
1. **Loss computation**: Always in FP32
2. **Gradient accumulation**: Always in FP32
3. **Optimizer updates**: Always in FP32 (master weights)
4. **Loss scaling**: Dynamic scaling for FP16 (not needed for BF16)

### FP16 vs BF16 Comparison
| Feature | FP16 | BF16 |
|---------|------|------|
| Exponent bits | 5 | 8 (same as FP32) |
| Mantissa bits | 10 | 7 |
| Dynamic range | Limited | Same as FP32 |
| Stability | Needs loss scaling | More stable |
| Hardware support | Broader | Newer GPUs |

### Tensor Core Requirements
- NVIDIA: Ampere (RTX 30xx), Ada (RTX 40xx), or newer
- AMD: MI200 series (CDNA 2) for matrix cores
- Matrix dimensions must be multiples of 8 (FP16) or 16 (BF16)

### API Example
```cpp
// Create mixed precision layer
MixedPrecisionGPULoRALayer layer(
    768, 768, 8, 1.0f,
    Device::cuda(),
    DType::FLOAT16  // Compute in FP16
);

// AMP training context
AMPContext amp(initial_scale=65536.0f);

// Training loop
for (auto& [input, target] : train_data) {
    // Forward in FP16
    auto output = layer.forward(input);
    
    // Loss in FP32
    float loss = mse_loss(output.to_fp32(), target);
    
    // Scale loss for FP16 backward
    loss = amp.scale_loss(loss);
    
    // Backward in FP16
    layer.backward(grad_loss);
    
    // Unscale gradients
    amp.unscale_gradients(layer.parameters());
    
    // Update in FP32 (master weights)
    optimizer.step();
    
    // Update loss scale
    amp.update_scale();
}
```

---

**Status**: Not Started  
**Estimated Completion**: Week 11 of Phase 10  
**Expected Benefit**: 2x speedup + 2x memory savings (enables larger models)
